### PR TITLE
Support mat_type dense

### PR DESCRIPTION
**Pull request by pefarrell**
Depends on https://github.com/OP2/PyOP2/pull/578 .

### CODE REVIEW

File: firedrake/matrix.py [hunk 1]
```diff
@@ -96,6 +96,7 @@ def __init__(self, a, bcs, mat_type, *args, **kwargs):
         # sets self._a, self._bcs, and self._mat_type
         super(Matrix, self).__init__(a, bcs, mat_type)
         options_prefix = kwargs.pop("options_prefix")
+        kwargs["mat_type"] = mat_type
         self.M = op2.Mat(*args, **kwargs)
```
Comment (wence-) [inline firedrake/matrix.py:100]: You can just say `op2.Mat(*args, mat_type=mat_type, **kwargs)`

Comment (pefarrell) [reply firedrake/matrix.py]: Great, thanks!

**Comment by wence-** [comment]
Tests? Projections would be fine, but preferably one scalar, one vector, one mixed.

**Comment by pefarrell** [comment]
I decided to add a mixed case to the test to be sure to be sure.

File: tests/regression/test_mat_type_dense.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    sp = {"snes_type": "ksponly",
+          "ksp_type": "preonly",
+          "pc_type": "lu",
+          "mat_type": mat_type}
+
+    solve(F == 0, u, bcs, solver_parameters=sp)
+    return u
+
+
+@pytest.mark.parametrize(('kind',), [('vector',), ('mixed',)])
```
Comment (wence-) [inline tests/regression/test_mat_type_dense.py:33]: FWIW for future reference, I think you can write:
```
@pytest.mark.parametrize("kind", ["vector", "mixed"])
```

instead here.

Comment (pefarrell) [reply tests/regression/test_mat_type_dense.py:33]: Ah, I was cargo-culting from other tests. Thanks!

**Comment by wence-** [comment]
Thanks.

**Pull request merged**